### PR TITLE
workarounds for browserify

### LINKED ISSUE #24: Browser support
Trying to get this working in the browser, and then later [hihat](https://github.com/Jam3/hihat) for debugging/breakpoints/etc.

After browserifying the demo in the readme, I get the following:

`Uncaught ReferenceError: setImmediate is not defined`

[relevant line](https://github.com/sindresorhus/ava/blob/97add1fcec62d510d72f1d26e1ea5ca972c495c6/index.js#L65)

(weird cause I thought browserify would have shimmed this)

### PR DESCRIPTION
**Pull request by danielepolencic**
fixes #24 

- if `stderr` is not available, inject the console in `squeak`
- `process.exit` is not available in the browser. Don't call the method if it is not available.

### CODE REVIEW

**Comment by sindresorhus** [comment]
I still think this is better fixed in `squeak`: https://github.com/sindresorhus/ava/issues/24#issuecomment-140687957

File: index.js [hunk 1]
```diff
@@ -83,7 +84,9 @@ function exit() {
 		stack(results);
 	}
 
-	process.exit(stats.failCount > 0 ? 1 : 0);
+	if (process.exit) {
```
Comment (sindresorhus) [inline index.js]: How will this work with `hihat`. According to its docs:

> The process will stay open until you call window.close()

Wouldn't it be better to call `window.close` here if in the `browser`. Or even better, keep it as before, and have `hihat` shim `process.exit` => https://github.com/Jam3/hihat/issues/27

**Comment by sindresorhus** [comment]
// @mattdesl

**Comment by mattdesl** [comment]
Hmm.. `window.close` would also work with [smokestack](https://github.com/hughsk/smokestack) but it seems like a pretty big assumption. 

In `hihat` the goal is interactive testing (fast save -> reload workflow). If `ava` exits the process on fail, it would break this development cycle. Maybe `ava` should not quit at all in the browser?

**Comment by sindresorhus** [comment]
@mattdesl Good point. So what should we do here?

**Comment by Qix-** [comment]
Hmm. Perhaps `ava.on('exit', ...)`, where ava detects if it's in node and registers the default handler of `process.exit()`? Let test implementers decide; we can add sensible defaults as necessary.

**Comment by Qix-** [comment]
Or even shim `process.exit()`? It fires the `exit` event (`process.on('exit', ...)`) so that might even be the ideal way anyway to keep full compatibility.

**Comment by vdemedes** [comment]
Agree with @mattdesl, on client-side `process.exit()` should do nothing or send an `exit` event, but definitely not close a window.

**Comment by mattdesl** [comment]
It seems sorta weird that an API would ever call `process.exit`. I feel like this should be left up to the user.

I feel like only the CLI should exit and handle errors itself, since in that case the user isn't handling them.

**Comment by vdemedes** [comment]
@mattdesl this happens, because we want to cover a case when test is being executed directly.

```
$ node test-something.js
```

So if only CLI would use `process.exit()`, in the above case process would exit always with zero (success).

**Comment by mattdesl** [comment]
I see. It's a bit of a special case then.

**Comment by arthurvr** [comment]
> Or even shim process.exit()? It fires the exit event (process.on('exit', ...)) so that might even be the ideal way anyway to keep full compatibility.

To me that sounds the most reasonable way to me. Though I guess this is kind of a rare case anyways.

**Comment by Qix-** [comment]
It is, but it's still probably the *correct* way. Just a thought.

**Comment by scottcorgan** [comment]
Just curious if there's been any progress for the browser. Only thing keeping me from Ava.

**Comment by sindresorhus** [comment]
@kevva said he would make `squeak` browserifyable and this PR needs to be updated to only include the exit change.

**Comment by sindresorhus** [comment]
Might be more, though, as we recently introduced parallism, which forks the processes. Not sure whether browserify handles that or if we need to do some special casing. Help welcome. Our main focus is elsewhere.

**Comment by kevva** [comment]
Should be fixed in https://github.com/kevva/squeak/commit/fd7333e343b1ff5f73cd14736ccc049c45dead3b.

**Comment by kevva** [comment]
> ... this PR needs to be updated to only include the exit change.

So what should we decide on? Just check for `process.exit` and do nothing if it doesn't exist?

**Comment by sindresorhus** [comment]
Closing in favor of #97.

**Pull request closed without merging**